File: process_icon_corners.py
#!/usr/bin/env python3
"""
图标圆角处理脚本
对所有图标文件进行统一的圆角裁切处理

使用方法:
    python process_icon_corners.py          # 处理所有图标
    python process_icon_corners.py --restore # 恢复原始图标
    
依赖:
    pip install Pillow
"""

import os
import sys
from PIL import Image, ImageDraw
import shutil

# 圆角半径（相对于图标尺寸的比例）
CORNER_RADIUS_RATIO = 0.22  # 22%的圆角，与CSS保持一致

def create_rounded_mask(size, radius):
    """创建圆角遮罩"""
    mask = Image.new('L', size, 0)
    draw = ImageDraw.Draw(mask)
    draw.rounded_rectangle([(0, 0), size], radius=radius, fill=255)
    return mask

def apply_rounded_corners(input_path, output_path):
    """对单个图标应用圆角"""
    try:
        print(f"🔄 处理: {os.path.basename(input_path)}")
        
        # 打开图像
        with Image.open(input_path) as img:
            # 确保是RGBA模式以支持透明度
            if img.mode != 'RGBA':
                img = img.convert('RGBA')
            
            # 获取图像尺寸
            width, height = img.size
            
            # 计算圆角半径
            radius = int(min(width, height) * CORNER_RADIUS_RATIO)
            
            # 创建圆角遮罩
            mask = create_rounded_mask((width, height), radius)
            
            # 创建输出图像（透明背景）
            output = Image.new('RGBA', (width, height), (0, 0, 0, 0))
            
            # 将原图像粘贴到输出图像
            output.paste(img, (0, 0))
            
            # 应用圆角遮罩
            output.putalpha(mask)
            
            # 保存处理后的图像
            output.save(output_path, 'PNG', optimize=True)
            
        print(f"✅ 完成: {os.path.basename(input_path)}")
        return True
        
    except Exception as e:
        print(f"❌ 失败: {os.path.basename(input_path)} - {str(e)}")
        return False

def process_all_icons():
    """批量处理所有图标"""
    source_dir = 'public/icon_style'
    backup_dir = 'public/icon_style_backup'
    
    if not os.path.exists(source_dir):
        print(f"❌ 源目录不存在: {source_dir}")
        return
    
    # 创建备份目录
    if not os.path.exists(backup_dir):
        os.makedirs(backup_dir)
        print("📁 创建备份目录")
    
    # 获取所有PNG文件
    files = [f for f in os.listdir(source_dir) if f.endswith('.png')]
    
    if not files:
        print("❌ 未找到PNG文件")
        return
    
    print(f"🔄 开始处理 {len(files)} 个图标文件...")
    print(f"📦 圆角半径: {int(CORNER_RADIUS_RATIO * 100)}%\n")
    
    success_count = 0
    fail_count = 0
    
    for file in files:
        input_path = os.path.join(source_dir, file)
        backup_path = os.path.join(backup_dir, file)
        temp_path = os.path.join(source_dir, f"temp_{file}")
        
        try:
            # 备份原文件
            if not os.path.exists(backup_path):
                shutil.copy2(input_path, backup_path)
            
            # 处理图像到临时文件
            success = apply_rounded_corners(input_path, temp_path)
            
            if success:
                # 替换原文件
                shutil.move(temp_path, input_path)
                success_count += 1
            else:
                # 清理临时文件
                if os.path.exists(temp_path):
                    os.remove(temp_path)
                fail_count += 1
                
        except Exception as e:
            print(f"处理 {file} 时出错: {str(e)}")
            fail_count += 1
            
            # 清理临时文件
            if os.path.exists(temp_path):
                os.remove(temp_path)
    
    print(f"\n📊 处理结果:")
    print(f"✅ 成功: {success_count} 个文件")
    print(f"❌ 失败: {fail_count} 个文件")
    
    if success_count > 0:
        print(f"\n💡 处理完成！")
        print(f"📁 原始文件已备份到: {backup_dir}")
        print("🔄 如需恢复原始文件，请运行: python process_icon_corners.py --restore")

def restore_original_icons():
    """恢复原始图标"""
    source_dir = 'public/icon_style'
    backup_dir = 'public/icon_style_backup'
    
    if not os.path.exists(backup_dir):
        print("❌ 未找到备份文件")
        return
    
    files = [f for f in os.listdir(backup_dir) if f.endswith('.png')]
    
    print(f"🔄 恢复 {len(files)} 个原始图标...")
    
    for file in files:
        backup_path = os.path.join(backup_dir, file)
        target_path = os.path.join(source_dir, file)
        
        try:
            shutil.copy2(backup_path, target_path)
            print(f"✅ 恢复: {file}")
        except Exception as e:
            print(f"❌ 恢复失败: {file} - {str(e)}")
    
    print("✅ 恢复完成！")

def check_dependencies():
    """检查依赖"""
    try:
        import PIL
        return True
    except ImportError:
        return False

def main():
    print("🖼️  图标圆角处理工具")
    print("=" * 40)
    
    # 检查依赖
    if not check_dependencies():
        print("❌ 未安装 Pillow 库")
        print("请运行: pip install Pillow")
        print("\n安装完成后重新运行此脚本")
        return
    
    print("✅ 依赖检查通过")
    
    # 检查参数
    if len(sys.argv) > 1 and sys.argv[1] == '--restore':
        restore_original_icons()
        return
    
    # 确认处理
    print(f"\n📦 圆角设置: {int(CORNER_RADIUS_RATIO * 100)}%")
    print("📁 源目录: public/icon_style/")
    print("📁 备份目录: public/icon_style_backup/")
    
    response = input("\n是否开始处理? (y/N): ").strip().lower()
    if response not in ['y', 'yes']:
        print("❌ 已取消处理")
        return
    
    try:
        process_all_icons()
    except KeyboardInterrupt:
        print("\n\n⚠️  处理被用户中断")
    except Exception as e:
        print(f"\n❌ 处理过程中出现错误: {str(e)}")
        print("请检查文件权限和目录结构")

if __name__ == "__main__":
    main()

File: scripts/process-icon-corners.js
/**
 * 图标圆角处理脚本
 * 对所有图标文件进行统一的圆角裁切处理
 */

import fs from 'fs';
import path from 'path';

// 圆角半径（相对于图标尺寸的比例）
const CORNER_RADIUS_RATIO = 0.22; // 22%的圆角，与CSS保持一致

/**
 * 使用Sharp库处理图像（如果可用）
 */
async function processWithSharp(inputPath, outputPath) {
    try {
        const sharp = await import('sharp');
        const image = sharp.default(inputPath);
        const { width, height } = await image.metadata();
        
        if (!width || !height) {
            throw new Error('无法获取图像尺寸');
        }
        
        const radius = Math.min(width, height) * CORNER_RADIUS_RATIO;
        
        // 创建圆角遮罩SVG
        const roundedCornerSvg = `
            <svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">
                <defs>
                    <clipPath id="rounded">
                        <rect width="${width}" height="${height}" rx="${radius}" ry="${radius}"/>
                    </clipPath>
                </defs>
                <rect width="${width}" height="${height}" fill="white" clip-path="url(#rounded)"/>
            </svg>
        `;
        
        await image
            .composite([{
                input: Buffer.from(roundedCornerSvg),
                blend: 'dest-in'
            }])
            .png()
            .toFile(outputPath);
            
        return true;
    } catch (error) {
        console.error(`Sharp处理失败: ${error.message}`);
        return false;
    }
}

/**
 * 使用Canvas处理图像（备用方案）
 */
async function processWithCanvas(inputPath, outputPath) {
    try {
        const { createCanvas, loadImage } = await import('canvas');
        
        const image = await loadImage(inputPath);
        const { width, height } = image;
        
        // 计算圆角半径
        const radius = Math.min(width, height) * CORNER_RADIUS_RATIO;
        
        // 创建输出画布
        const canvas = createCanvas(width, height);
        const ctx = canvas.getContext('2d');
        
        // 创建圆角路径
        ctx.beginPath();
        ctx.moveTo(radius, 0);
        ctx.lineTo(width - radius, 0);
        ctx.quadraticCurveTo(width, 0, width, radius);
        ctx.lineTo(width, height - radius);
        ctx.quadraticCurveTo(width, height, width - radius, height);
        ctx.lineTo(radius, height);
        ctx.quadraticCurveTo(0, height, 0, height - radius);
        ctx.lineTo(0, radius);
        ctx.quadraticCurveTo(0, 0, radius, 0);
        ctx.closePath();
        ctx.clip();
        
        // 绘制图像
        ctx.drawImage(image, 0, 0, width, height);
        
        // 保存处理后的图像
        const buffer = canvas.toBuffer('image/png');
        fs.writeFileSync(outputPath, buffer);
        
        return true;
    } catch (error) {
        console.error(`Canvas处理失败: ${error.message}`);
        return false;
    }
}

/**
 * 对单个图标应用圆角
 */
async function applyRoundedCorners(inputPath, outputPath) {
    try {
        console.log(`🔄 处理: ${path.basename(inputPath)}`);
        
        // 优先尝试Sharp，然后是Canvas
        let success = await processWithSharp(inputPath, outputPath);
        
        if (!success) {
            success = await processWithCanvas(inputPath, outputPath);
        }
        
        if (success) {
            console.log(`✅ 完成: ${path.basename(inputPath)}`);
            return true;
        } else {
            console.log(`❌ 失败: ${path.basename(inputPath)}`);
            return false;
        }
    } catch (error) {
        console.error(`❌ 处理失败: ${path.basename(inputPath)} - ${error.message}`);
        return false;
    }
}

/**
 * 批量处理所有图标
 */
async function processAllIcons() {
    const sourceDir = 'public/icon_style';
    const backupDir = 'public/icon_style_backup';
    
    // 创建备份目录
    if (!fs.existsSync(backupDir)) {
        fs.mkdirSync(backupDir, { recursive: true });
        console.log('📁 创建备份目录');
    }
    
    // 获取所有PNG文件
    const files = fs.readdirSync(sourceDir).filter(file => file.endsWith('.png'));
    
    console.log(`🔄 开始处理 ${files.length} 个图标文件...`);
    console.log(`📦 圆角半径: ${(CORNER_RADIUS_RATIO * 100).toFixed(0)}%\n`);
    
    let successCount = 0;
    let failCount = 0;
    
    for (const file of files) {
        const inputPath = path.join(sourceDir, file);
        const backupPath = path.join(backupDir, file);
        const tempPath = path.join(sourceDir, `temp_${file}`);
        
        try {
            // 备份原文件
            if (!fs.existsSync(backupPath)) {
                fs.copyFileSync(inputPath, backupPath);
            }
            
            // 处理图像到临时文件
            const success = await applyRoundedCorners(inputPath, tempPath);
            
            if (success) {
                // 替换原文件
                fs.renameSync(tempPath, inputPath);
                successCount++;
            } else {
                // 清理临时文件
                if (fs.existsSync(tempPath)) {
                    fs.unlinkSync(tempPath);
                }
                failCount++;
            }
        } catch (error) {
            console.error(`处理 ${file} 时出错:`, error.message);
            failCount++;
            
            // 清理临时文件
            if (fs.existsSync(tempPath)) {
                fs.unlinkSync(tempPath);
            }
        }
    }
    
    console.log('\n📊 处理结果:');
    console.log(`✅ 成功: ${successCount} 个文件`);
    console.log(`❌ 失败: ${failCount} 个文件`);
    
    if (successCount > 0) {
        console.log('\n💡 处理完成！');
        console.log(`📁 原始文件已备份到: ${backupDir}`);
        console.log('🔄 如需恢复原始文件，请运行: npm run restore-icons');
    }
}

/**
 * 恢复原始图标
 */
function restoreOriginalIcons() {
    const sourceDir = 'public/icon_style';
    const backupDir = 'public/icon_style_backup';
    
    if (!fs.existsSync(backupDir)) {
        console.log('❌ 未找到备份文件');
        return;
    }
    
    const files = fs.readdirSync(backupDir).filter(file => file.endsWith('.png'));
    
    console.log(`🔄 恢复 ${files.length} 个原始图标...`);
    
    for (const file of files) {
        const backupPath = path.join(backupDir, file);
        const targetPath = path.join(sourceDir, file);
        
        try {
            fs.copyFileSync(backupPath, targetPath);
            console.log(`✅ 恢复: ${file}`);
        } catch (error) {
            console.log(`❌ 恢复失败: ${file} - ${error.message}`);
        }
    }
    
    console.log('✅ 恢复完成！');
}

/**
 * 检查依赖
 */
async function checkDependencies() {
    const deps = [];
    
    try {
        await import('sharp');
        deps.push('sharp');
    } catch (e) {
        // Sharp不可用
    }
    
    try {
        await import('canvas');
        deps.push('canvas');
    } catch (e) {
        // Canvas不可用
    }
    
    return deps;
}

// 主函数
async function main() {
    console.log('🖼️  图标圆角处理工具\n');
    
    const args = process.argv.slice(2);
    
    if (args.includes('--restore')) {
        restoreOriginalIcons();
        return;
    }
    
    const availableDeps = await checkDependencies();
    
    if (availableDeps.length === 0) {
        console.log('❌ 未找到图像处理依赖');
        console.log('请安装以下依赖之一:');
        console.log('  npm install sharp  (推荐，性能更好)');
        console.log('  npm install canvas  (备用方案)');
        return;
    }
    
    console.log(`📦 可用的图像处理库: ${availableDeps.join(', ')}`);
    
    try {
        await processAllIcons();
    } catch (error) {
        console.error('❌ 处理过程中出现错误:', error.message);
    }
}

main();